### ISSUE TITLE
4.1.1rc3: Header tests all negative

### ISSUE THREAD

**Issue by pkovacs** [openpmix/openpmix#2333]
Something happened with the header checks as they are coming up negative now.  Perhaps something in `config\pmix.m4` ?

```
============================================================================
== Header file tests
============================================================================
checking if gcc supports (-Werror)... yes
checking for arpa/inet.h... no
checking for fcntl.h... no
checking for ifaddrs.h... no
checking for inttypes.h... (cached) yes
checking for libgen.h... no
checking for net/uio.h... no
checking for netinet/in.h... (cached) yes
checking for stdint.h... (cached) yes
checking for stddef.h... no
checking for stdlib.h... (cached) yes
checking for string.h... (cached) yes
checking for strings.h... (cached) yes
checking for sys/ioctl.h... no
checking for sys/param.h... no
checking for sys/select.h... no
checking for sys/socket.h... no
checking for sys/sockio.h... no
checking for stdarg.h... no
checking for sys/stat.h... (cached) yes
checking for sys/time.h... no
checking for sys/types.h... (cached) yes
checking for sys/un.h... no
checking for sys/uio.h... no
checking for sys/wait.h... no
checking for syslog.h... no
checking for time.h... no
checking for unistd.h... (cached) yes
checking for dirent.h... no
checking for crt_externs.h... no
checking for signal.h... no
checking for ioLib.h... no
checking for sockLib.h... no
checking for hostLib.h... no
checking for limits.h... no
checking for sys/fcntl.h... no
checking for sys/statfs.h... no
checking for sys/statvfs.h... no
checking for netdb.h... (cached) yes
checking for ucred.h... no
checking for zlib.h... no
checking for sys/auxv.h... no
checking for sys/sysctl.h... no
checking for termio.h... no
checking for termios.h... no
checking for pty.h... no
checking for libutil.h... no
checking for util.h... no
checking for grp.h... no
checking for sys/cdefs.h... no
```

**Comment by rhc54** [comment]
Odd - can you see what the warning is on one of those tests?

**Comment by rhc54** [comment]
Oh, and what compiler is this?

**Comment by pkovacs** [comment]
Rather it's not the presence of that flag -- it's the absence of it.

**Comment by rhc54** [comment]
So just to be clear: are you saying that Fedora adds that flag (I imagine by defining it in `CFLAGS` prior to running configure), and we somehow remove it prior to running the checks?

**Comment by pkovacs** [comment]
The issue is that `-Werror` was adding here 55a36465f5d015e61484f3930b990d0d487e4ddf which has the side effect above.  The fortify source flag is added by the rpm enviroment at compile time but not as early as autogen.pl/configure time, thus the conftests don't have the flag and -Werror fails the tests.  I am working around it by adding the needed definition for autogen/configure and then removing it for the compile since it gets added automatically at the compile phase.  Essentially this:

```
%build
# add needed flag for -Werror constraint
export CPPFLAGS="-D_FORTIFY_SOURCE=2"
%{_builddir}/%{name_version}/autogen.pl
%configure \
    --prefix=%{_prefix} \
    --sysconfdir=%{_sysconfdir}/%{name} \
    --disable-static \
    --disable-silent-rules \
    --enable-shared \
    --with-munge

# remove manual flag since we it is added anyway 
export CPPFLAGS=""
%make_build
```

**Comment by rhc54** [comment]
Remind me: you are building from the tarball, yes? Not a git clone of the repo branch?

**Comment by rhc54** [comment]
Before you spend a bunch of time on this, let me try and create a patch that will hopefully fix these problems and still get thru Clang-10

**Comment by pkovacs** [comment]
I'm building from the 4.1.1rc3.tar.bz2 tarball, yes.  I tried with `--disable-picky` but both classes of errors persist -- the headers fail due to the `-Werror` addition and the linker is barking about -fPIE as I mentioned above.   If I patch out the `-Werror` (above) the headers are ok, but the PIE problem persists.  I suspect something new might be removing, filtering or otherwise overriding cflags.   On my (Red Hat/Fedora) side, all code is compiled position independent (-fPIC) and all binaries are compiled (-fPIE) to allow for random placement in memory as the programs load.

**Comment by pkovacs** [comment]
This is part of the macros file we have for RH/Fedora.  Every rpm-based build would normally see these cppflags below.  Something new in pmix is clobbering it.  Not sure yet on the PIE problem.

```
# /usr/lib/rpm/redhat/macros
%_preprocessor_defines  -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS
```

**Comment by rhc54** [comment]
Can you please try the following patch:
```diff
diff --git a/config/pmix.m4 b/config/pmix.m4
index 59597371..2cd21b83 100644
--- a/config/pmix.m4
+++ b/config/pmix.m4
@@ -431,10 +431,6 @@ AC_DEFUN([PMIX_SETUP_CORE],[
 
     pmix_show_title "Header file tests"
 
-    PMIX_VAR_SCOPE_PUSH(PMIX_CFLAGS_save_for_headers)
-    PMIX_CFLAGS_save_for_headers=$CFLAGS
-    _PMIX_CHECK_SPECIFIC_CFLAGS(-Werror, Werror)
-
     AC_CHECK_HEADERS([arpa/inet.h \
                       fcntl.h ifaddrs.h inttypes.h libgen.h \
                       net/uio.h netinet/in.h \
@@ -514,9 +510,6 @@ AC_DEFUN([PMIX_SETUP_CORE],[
                        [Whether to use <stdbool.h> or not])
     AC_MSG_RESULT([$MSG])
 
-    CFLAGS=$PMIX_CFLAGS_save_for_headers
-    PMIX_VAR_SCOPE_POP
-
     # checkpoint results
     AC_CACHE_SAVE
 
@@ -1075,23 +1068,16 @@ fi
 #
 
 AC_MSG_CHECKING([if want developer-level compiler pickyness])
-AC_ARG_ENABLE(picky,
-    AS_HELP_STRING([--enable-picky],
+AC_ARG_ENABLE(devel-check,
+    AS_HELP_STRING([--enable-devel-check],
                    [enable developer-level compiler pickyness when building PMIx (default: disabled)]))
-if test "$enable_picky" = "yes"; then
+if test "$enable_devel_check" = "yes"; then
     AC_MSG_RESULT([yes])
     WANT_PICKY_COMPILER=1
 else
     AC_MSG_RESULT([no])
     WANT_PICKY_COMPILER=0
 fi
-#################### Early development override ####################
-#if test "$WANT_PICKY_COMPILER" = "0" && test -z "$enable_picky" && test "$PMIX_DEVEL" = "1"; then
-#    WANT_PICKY_COMPILER=1
-#    echo "--> developer override: enable picky compiler by default"
-#fi
-#################### Early development override ####################
-
 AC_DEFINE_UNQUOTED(PMIX_PICKY_COMPILERS, $WANT_PICKY_COMPILER,
                    [Whether or not we are using picky compiler settings])
 
diff --git a/test/Makefile.am b/test/Makefile.am
index 7d920e7a..d864d8e7 100644
--- a/test/Makefile.am
+++ b/test/Makefile.am
@@ -39,7 +39,6 @@ headers = test_common.h cli_stages.h server_callbacks.h utils.h test_fence.h \
 
 AM_CPPFLAGS = -I$(top_builddir)/src -I$(top_builddir)/src/include -I$(top_builddir)/src/api
 # we do NOT want picky compilers down here
-CFLAGS = $(PMIX_CFLAGS_BEFORE_PICKY)
 
 noinst_SCRIPTS = pmix_client_otheruser.sh \
        run_tests00.pl \
diff --git a/test/simple/Makefile.am b/test/simple/Makefile.am
index 1b17fa95..cc5df268 100644
--- a/test/simple/Makefile.am
+++ b/test/simple/Makefile.am
@@ -22,7 +22,6 @@
 
 AM_CPPFLAGS = -I$(top_builddir)/src -I$(top_builddir)/src/include -I$(top_builddir)/include -I$(top_builddir)/include/pmix
 # we do NOT want picky compilers down here
-CFLAGS = $(PMIX_CFLAGS_BEFORE_PICKY)
 
 headers = simptest.h
 
diff --git a/test/test_v2/Makefile.am b/test/test_v2/Makefile.am
index 873d89fa..a21c8231 100644
--- a/test/test_v2/Makefile.am
+++ b/test/test_v2/Makefile.am
@@ -27,8 +27,6 @@
 headers = cli_stages.h server_callbacks.h test_common.h test_server.h
 
 AM_CPPFLAGS = -I$(top_builddir)/src -I$(top_builddir)/src/include -I$(top_builddir)/src/api
-# we do NOT want picky compilers down here
-CFLAGS = $(PMIX_CFLAGS_BEFORE_PICKY)
 
 # Can be added in for purposes of debugging on platforms building with gcc (gcc-specific flags)
 #AM_CFLAGS = -fno-omit-frame-pointer -g -Wfatal-errors #-Wall -Wextra -Wpedantic -Wconversion -Wshadow
```
This still works around the Clang-10 issue - hopefully, will solve this one too.

**Comment by rhc54** [comment]
Really? Huh, that's odd. Let me try again

**Comment by rhc54** [comment]
Not sure what to say here - it applies okay for me:
```text
$ patch -p1 < pt.diff 
patching file config/pmix.m4
patching file test/Makefile.am
Hunk #1 succeeded at 39 with fuzz 1.
patching file test/simple/Makefile.am
patching file test/test_v2/Makefile.am
$
```
Since all we want to do is verify that it solves the problem, can you perhaps resolve the issue by hand and test the build?

**Comment by pkovacs** [comment]
Just for documentation and comparison, here is my clean pmix 4.1.0 build log -- exact same environment.

[build-4.1.0-1.fc36.log](https://github.com/openpmix/openpmix/files/7318654/build-4.1.0-1.fc36.log)

**Comment by rhc54** [comment]
Hmmm...ok, I think I found one more place I hadn't fixed. Try removing the following lines in the examples/Makefile.am file:
```diff
diff --git a/examples/Makefile.am b/examples/Makefile.am
index 1077dece..99ed14c8 100644
--- a/examples/Makefile.am
+++ b/examples/Makefile.am
@@ -22,8 +22,6 @@
 headers = examples.h
 
 AM_CPPFLAGS = -I$(top_builddir)/src -I$(top_builddir)/src/include -I$(top_builddir)/include -I$(top_builddir)/include/pmix
-# we do NOT want picky compilers down here
-CFLAGS = $(PMIX_CFLAGS_BEFORE_PICKY)
 
 noinst_PROGRAMS = client client2 dmodex dynamic fault pub pubi tool debugger debuggerd alloc jctrl group asyncgroup
 if !WANT_HIDDEN
```

**Comment by pkovacs** [comment]
Success!  That additional patch above allows the build to complete.   Log attached.

[build-4.1.1-0.1.rc3.fc36.log](https://github.com/openpmix/openpmix/files/7319368/build-4.1.1-0.1.rc3.fc36.log)

**Comment by rhc54** [comment]
Yeah, I posted a patch to update Slurm awhile back - pretty sure it was pulled in, but maybe not in time for their 21.08? Anyway, glad you were able to adjust.

I'm not sure who is handling PRTE, but I thought someone had picked it up. @opoplawski - do you happen to know? I ask because (a) people are using it fairly broadly now, and (b) it will be needed for OMPI v5.

**Comment by opoplawski** [comment]
I'm not aware of anyone picking up PRRTE (sp?) for Fedora yet.  I suppose it will have to end up being me, because I'm not sure who else is interested in openmpi work in Fedora these days.

**Comment by pkovacs** [comment]
I built and released pmix 4.1.1rc4 to Fedora rawhide (cutting edge).  No problems with the rc4 build.

**Comment by rhc54** [comment]
Thanks Phil!
